### DISCUSSION TITLE
First/Soft API for Attendance Project

### DISCUSSION THREAD

**Discussion by hongkevi** [MSU-AI/Attendance-Project#3] [Ideas]
![image](https://user-images.githubusercontent.com/78090720/155242571-74ce9093-c541-4448-be37-bda0b8bbb427.png)

First/Soft Version of API for the code for the attendance project.

2/22/2022

### 🔻Required Modules and Imports

These are the python modules / or any libraries needed in order to run the code as of 2/22/2022.

**Hand Interaction Modules**
- [ ]  [requirements.txt](https://github.com/MSU-AI/Attendance-Project/files/8121100/requirements.txt)
As a list: matplotlib, mediapipe, numpy, opencv2, pandas, scikit-learn

**Face Recognition Modules**
- [ ]  [requirements.txt](https://github.com/MSU-AI/Attendance-Project/files/8121126/requirements.txt)
 As a list: cmake, dlib wheel, face recognition
- [ ]  **DLIB WHEEL** - Not included in the text file. Link to dlib wheel: [Click](https://drive.google.com/drive/folders/15JzSg2aChlEc2ETgcJlz_RNAHzX_dKSe?usp=sharing)

**NOTE: Hand interaction ``requirement.txt`` is in the current repository. The Face recognition ``requirement.txt`` is not in the current repository. Please double-check before you download!

### 🔻Components, Functions, and Classes... Which files do we need to use?

In order to have the program start, certain steps are needed to prepare and have code run correctly.

- [ ]  All files in the current branch (interaction) are needed to run the program.
- [ ]  ``train.py`` program will need to be run **only** once if you are starting the program for the very first time.
This creates trained data of what hand is like and inputs it with the hand interaction.
- [ ]  *Not required but if testing for face indication, images needed* - Have image (jpg) files in the ``face_images`` file. The title of the pictures should indicate person's name. (Base64 - JPEG convert information on the third bullet point)
- [ ]  Functions/ Classes running while the code is being ran.
``recognition.py`` is the main code that runs everything to initialize and send create input/output camera frames to detect hand and face. Everything in ``recognition.py`` such as functions/classes are needed to run the program. ``simple_facerec.py`` is also needed when running the code because face recognition pulls functions/classes to initialize faces.

###  🔻Face Images - Formatting and File Type

Feedback from backend/frontend team needs to convert images.
We require jpg images, but any jpg should be fine in order to detect faces. There are some anomalies to test with jpg but it should be working as intended.

- [ ] A program to convert Base64 to JPG format is provided here: [Click](https://github.com/MSU-AI/Attendance-Project/discussions/2)
*credit to [Fanurs](https://github.com/Fanurs)

Additional inquiries will be needed in order to get this image converter automated!

###  🔻 Code Output / Code Return
 
As of the current version as of 2/22/2022, there is no attendance point system as we are working to implement it soon. The current code is providing feedback to the user if the face is recognized and if the hand is creating a sign that is recognized by the trained data inside the code.

So as a result of this, it is certain that only **frames of images** are being returned through the code. Visible feedback is sent through the frames if the hand or face is being recognized. If there are more questions to be asked, be sure to communicate with the coding team or me to answer more of your questions!

Code being returned:
```python
with OpencvCamera() as camera:
        while cv2.waitKey(1) != ord('q'):
            _, frame = camera.capture.read()
            frame = cv2.flip(frame, 1) # mirror

            hand_tracker.process_frame(frame)
            if hand_tracker.found_hand():
                df_hands = hand_tracker.get_all_hand_dataframes(frame)
                # hand_tracker.draw_all_hands(frame)

                for i_hand, df_hand in enumerate(df_hands):
                    label = labels[hand_tracker.predict(df_hand)]

                    # display prediction
                    lm = hand_tracker.solution_outputs.multi_hand_landmarks[i_hand].landmark
                    x_pixel = int(np.mean([lm[0].x, lm[9].x]) * frame.shape[1])
                    y_pixel = int(np.mean([lm[0].y, lm[9].y]) * frame.shape[0])
                    cv2.putText(
                        frame, label,
                        org=(x_pixel, y_pixel),
                        fontFace=cv2.FONT_HERSHEY_DUPLEX,
                        fontScale=0.5,
                        color=(0, 255, 0),
                    )
            hand_tracker.clear_solution_outputs()

            time.sleep(0.03)
            # Start of Face Recog
            #detect faces
            face_locations, face_names = sfr.detect_known_faces(frame)   
            
            #Draw rectangle and print names
            
            for face_loc, name in zip(face_locations, face_names):
                
                y1,x1,y2,x2 = face_loc[0],face_loc[1],face_loc[2],face_loc[3]
                
                cv2.putText(frame,name,(x1,y1 - 10),cv2.FONT_HERSHEY_DUPLEX,1,(255,0,0),2)
                
                cv2.rectangle(frame,(x1,y1),(x2,y2),(0,0,200),2)
            
            cv2.imshow('Frame',frame)
```

###  🔻 Conclusion

This API is not perfect just like any other things so far. But currently, it is working to be improved so it is understandable to run and understand what the code is doing! If there are any errors or uncertainty, please contact the coding team or edit this post to clear up any bugs/errors!

MSU AI Club - Coding Team

**Comment by fanurs** [comment]
@hongkevi Good post!
I'm just wondering if other teams have asked for explicit function inputs/outputs format they want? Currently, what we have still looks like a big chunk of "main" function. Ideally they may want one single function like `get_hand_gesture(frame)` that returns, for example, `"thumbs-up"`, and `get_face(frame) -> "Obama"`?
